composer/trainer/dist_strategy.py

```diff
@@ -0,0 +1,261 @@
+            parameters in the model that are not being trained.
+    """
+    if dist.is_available() and dist.is_initialized():
+        if any((p.requires_grad for p in module.parameters())):
+            log.debug('Wrapping model with DistributedDataParallel')
+            ddp_model = DistributedDataParallel(module, find_unused_parameters=find_unused_parameters)
+            return ddp_model
+        return module
+    if dist.is_available():
+        raise RuntimeError('Please call dist.initialize_dist() before calling ddp.prepare_module()')
+
+    raise RuntimeError('When the world size is > 1, ``torch.distributed`` must be used. However, it is '
+                       'not available in your installation of PyTorch. Please install or build PyTorch '
+                       'with distributed support.')
+
+
+def prepare_fsdp_module(model: torch.nn.Module, optimizers: Optional[Union[torch.optim.Optimizer,
+                                                                           Sequence[torch.optim.Optimizer]]],
+                        fsdp_config: Dict[str, Any], precision: Precision) -> None:
+    """Prepare a module (assumed ComposerModel) and optimizer for use with :class:`torch.distributed.fsdp.FullyShardedDataParallel`.
+
+    Args:
+        model (torch.nn.Module): The model to wrap.
+        optimizers (torch.optim.Optimizer | Sequence[torch.optim.Optimizer], optional): The optimizer for `model`, assumed to have a single param group := model.parameters().
+        fsdp_config (Dict[str, Any]): todo
+        precision: (Precision): todo
+    """
+    if version.parse(torch.__version__) < version.parse('1.12.0'):
+        raise RuntimeError('To use FSDP with Composer, you must use torch>=1.12.0.')
+    else:
+        from torch.distributed.fsdp import (BackwardPrefetch, CPUOffload, FullyShardedDataParallel, MixedPrecision,
+                                            ShardingStrategy)
+
+    sharding_map = {
+        'NO_SHARD': ShardingStrategy.NO_SHARD,
+        'SHARD_GRAD_OP': ShardingStrategy.SHARD_GRAD_OP,
+        'FULL_SHARD': ShardingStrategy.FULL_SHARD,
+    }
+    sharding_strategy = sharding_map[fsdp_config.get('sharding_strategy', 'FULL_SHARD').upper()]
+
+    cpu_offload = CPUOffload(offload_params=True) if fsdp_config.get('cpu_offload', False) else None
+    if cpu_offload is not None:
+        raise ValueError('FSDP CPU Offload not supported yet.')
+
+    def _get_torch_dtype(dtype: Union[Precision, str]):
+        dtype = dtype.value if isinstance(dtype, Precision) else dtype
+        if dtype in ['float32', 'torch.float32', 'fp32']:
+            return torch.float32
+        elif dtype in ['float16', 'torch.float16', 'half', 'fp16', 'amp', 'amp_fp16']:
+            return torch.float16
+        elif dtype in ['bfloat16', 'bfloat', 'torch.bfloat16', 'bf16', 'amp_bf16']:
+            return torch.bfloat16
+        else:
+            raise ValueError(f'Not sure how to convert dtype={dtype} to a torch dtype.')
```

Comment:
minor nit: can we move this outside the function? stylistically, I think we avoid nested functions as much as possible unless it bakes specific arguments in the construction